// redux-form-yup/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

var C__Code_Personal_reduxFormYup_node_modules_isNil = function (obj) {

  return obj == null;
};

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var C__Code_Personal_reduxFormYup_node_modules_isSymbol = createCommonjsModule(function (module) {

var toStr = Object.prototype.toString;
var hasSymbols = typeof Symbol === 'function' && typeof Symbol() === 'symbol';

if (hasSymbols) {
	var symToStr = Symbol.prototype.toString;
	var symStringRegex = /^Symbol\(.*\)$/;
	var isSymbolObject = function isSymbolObject(value) {
		if (typeof value.valueOf() !== 'symbol') { return false; }
		return symStringRegex.test(symToStr.call(value));
	};
	module.exports = function isSymbol(value) {
		if (typeof value === 'symbol') { return true; }
		if (toStr.call(value) !== '[object Symbol]') { return false; }
		try {
			return isSymbolObject(value);
		} catch (e) {
			return false;
		}
	};
} else {
	module.exports = function isSymbol(value) {
		// this environment does not support Symbols.
		return false;
	};
}
});

/**
 * isArray
 */

var isArray = Array.isArray;

/**
 * toString
 */

var str = Object.prototype.toString;

/**
 * Whether or not the given `val`
 * is an array.
 *
 * example:
 *
 *        isArray([]);
 *        // > true
 *        isArray(arguments);
 *        // > false
 *        isArray('');
 *        // > false
 *
 * @param {mixed} val
 * @return {bool}
 */

var C__Code_Personal_reduxFormYup_node_modules_isArray = isArray || function (val) {
  return !! val && '[object Array]' == str.call(val);
};

var C__Code_Personal_reduxFormYup_node_modules_isWindow = function (obj) {

  if (obj == null) {
    return false;
  }

  var o = Object(obj);

  return o === o.window;
};

var C__Code_Personal_reduxFormYup_node_modules_isFunction = isFunction;

var toString = Object.prototype.toString;

function isFunction (fn) {
  var string = toString.call(fn);
  return string === '[object Function]' ||
    (typeof fn === 'function' && string !== '[object RegExp]') ||
    (typeof window !== 'undefined' &&
     // IE8 and below
     (fn === window.setTimeout ||
      fn === window.alert ||
      fn === window.confirm ||
      fn === window.prompt))
}

var C__Code_Personal_reduxFormYup_node_modules_isArrayLike = function (obj) {

  if (!obj) {
    return false;
  }

  if (C__Code_Personal_reduxFormYup_node_modules_isArray(obj)) {
    return true;
  }

  if (C__Code_Personal_reduxFormYup_node_modules_isFunction(obj) || C__Code_Personal_reduxFormYup_node_modules_isWindow(obj)) {
    return false;
  }

  obj = Object(obj);

  var length = 'length' in obj && obj.length;

  if (obj.nodeType === 1 && length) {
    return true;
  }

  return length === 0 ||
    typeof length === 'number' && length > 0 && ( length - 1 ) in obj;
};

var C__Code_Personal_reduxFormYup_node_modules_isObject = function isObject(x) {
	return typeof x === "object" && x !== null;
};

/* global Symbol */






var C__Code_Personal_reduxFormYup_node_modules_toStr = function (value) {

  if (typeof value === 'string') {
    return value;
  }

  if (C__Code_Personal_reduxFormYup_node_modules_isNil(value)) {
    return '';
  }

  if (C__Code_Personal_reduxFormYup_node_modules_isSymbol(value)) {
    return Symbol.prototype.toString.call(value);
  }

  if (C__Code_Personal_reduxFormYup_node_modules_isObject(value) && C__Code_Personal_reduxFormYup_node_modules_isFunction(value.toString)) {
    return value.toString();
  }

  var result = '' + value;

  return (result === '0' && (1 / value) === -1 / 0) ? '-0' : result;
};

var rePropName   = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]/g;
var reEscapeChar = /\\(\\)?/g;


function toKey(key) {
  return (typeof key === 'string' || C__Code_Personal_reduxFormYup_node_modules_isSymbol(key))
    ? key
    : ('' + key);
}


var C__Code_Personal_reduxFormYup_node_modules_toPath = function (value) {

  if (C__Code_Personal_reduxFormYup_node_modules_isNil(value)) {
    return [];
  }

  if (C__Code_Personal_reduxFormYup_node_modules_isSymbol(value)) {
    return [value];
  }


  var result = [];

  if (typeof value !== 'string' && C__Code_Personal_reduxFormYup_node_modules_isArrayLike(value)) {
    for (var i = 0, l = value.length; i < l; i++) {
      result.push(toKey(value[i]));
    }
    return result;
  }


  C__Code_Personal_reduxFormYup_node_modules_toStr(value)
    .replace(rePropName, function (match, number, quote, string) {
      result.push(quote ? string.replace(reEscapeChar, '$1') : (number || match));
    });

  return result;
};

var MAX_SAFE_INTEGER = 9007199254740991;

var C__Code_Personal_reduxFormYup_node_modules_isIndex = function (value, length) {

  value  = (typeof value === 'number' || /^(?:0|[1-9]\d*)$/.test(value)) ? +value : -1;
  length = C__Code_Personal_reduxFormYup_node_modules_isNil(length) ? MAX_SAFE_INTEGER : length;

  return value > -1 && value % 1 === 0 && value < length;
};

function assignValue(object, key, value) {

  var objValue = object[key];

  if (!(Object.prototype.hasOwnProperty.call(object, key) && objValue === value) ||
    (value === undefined && !(key in object))) {

    object[key] = value;
  }
}


var C__Code_Personal_reduxFormYup_node_modules_objectSet = function (object, path, value) {

  if (C__Code_Personal_reduxFormYup_node_modules_isNil(object)) {
    return object;
  }

  var nested    = object;
  var paths     = C__Code_Personal_reduxFormYup_node_modules_toPath(path);
  var index     = -1;
  var length    = paths.length;
  var lastIndex = length - 1;

  while (!C__Code_Personal_reduxFormYup_node_modules_isNil(nested) && ++index < length) {

    var key = paths[index];

    if (C__Code_Personal_reduxFormYup_node_modules_isObject(nested)) {

      var newValue = value;

      if (index !== lastIndex) {

        var objValue = nested[key];

        newValue = C__Code_Personal_reduxFormYup_node_modules_isNil(objValue)
          ? (C__Code_Personal_reduxFormYup_node_modules_isIndex(paths[index + 1]) ? [] : {})
          : objValue;
      }

      assignValue(nested, key, newValue);
    }

    nested = nested[key];
  }

  return object;
};

var _this = undefined;
var asyncValidate = function (schema) {
    return function (values) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, schema
                        .validate(values, { abortEarly: false })
                        .then(function () { })
                        .catch(function (errors) {
                        var formErrors = {};
                        errors.inner.forEach(function (error) {
                            C__Code_Personal_reduxFormYup_node_modules_objectSet(formErrors, error.path, error.message);
                        });
                        return formErrors;
                    })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
};

var shouldAsyncValidate = function () { return true; };

exports.asyncValidate = asyncValidate;
exports.shouldAsyncValidate = shouldAsyncValidate;
